server/core/src/main/java/dev/slimevr/firmware/OTAUpdateTask.kt

```diff
@@ -0,0 +1,180 @@
+				val resultText = "$hashedPassword:$authToken:$signature"
+				val payload = bytesToMd5(resultText.toByteArray())
+
+				val authMessage = "$AUTH $signature $payload\n"
+
+				socket.soTimeout = 10000
+				socket.send(
+					DatagramPacket(
+						authMessage.toByteArray(),
+						authMessage.length,
+						deviceIp,
+						PORT,
+					),
+				)
+
+				val authResponsePacket = DatagramPacket(receiveBuffer, receiveBuffer.size)
+				socket.receive(authResponsePacket)
+
+				val authResponse = String(authResponsePacket.data, 0, authResponsePacket.length)
+
+				return authResponse == "OK"
+			}
+		} catch (e: Exception) {
+			return false
+		}
+	}
+
+	private fun upload(serverSocket: ServerSocket): Boolean {
+		try {
+			LogManager.info("[OTAUpdate] Starting on: ${serverSocket.localPort}")
+			LogManager.info("[OTAUpdate] Waiting for device...")
+
+			val connection = serverSocket.accept()
+			connection.setSoTimeout(1000)
+
+			val dos = DataOutputStream(connection.getOutputStream())
+			val dis = DataInputStream(connection.getInputStream())
+
+			LogManager.info("[OTAUpdate] Upload size: ${firmware.size} bytes")
+			var offset = 0
+			val chunkSize = 2048
+			while (offset != firmware.size) {
+				statusCallback.accept(
+					UpdateStatusEvent(
+						deviceId,
+						FirmwareUpdateStatus.UPLOADING,
+						((offset.toDouble() / firmware.size) * 100).toInt(),
+					),
+				)
+
+				val chunkLen = min(chunkSize, (firmware.size - offset))
+				dos.write(firmware, offset, chunkLen)
+				dos.flush()
+				offset += chunkLen
+
+				// Those skipped bytes are the size written to the MCU. We do not really need that information,
+				// so we simply skip it.
+				// The reason those bytes are skipped here is to not have to skip all of them when checking
+				// for the OK response. Saving time
+				dis.skipNBytes(4)
+			}
+
+			LogManager.info("[OTAUpdate] Waiting for result...")
+			// We set the timeout of the connection bigger as it can take some time for the MCU
+			// to confirm that everything is ok
+			connection.setSoTimeout(10000)
+			val responseBytes = dis.readAllBytes()
+			val response = String(responseBytes)
+
+			return response.contains("OK")
+		} catch (e: Exception) {
+			LogManager.severe("Unable to upload the firmware using ota", e)
+			return false
+		}
+	}
+
+	fun run() {
+		ServerSocket(0).use { serverSocket ->
+			if (!authenticate(serverSocket.localPort)) {
+				statusCallback.accept(
+					UpdateStatusEvent(
+						deviceId,
+						FirmwareUpdateStatus.ERROR_AUTHENTICATION_FAILED,
+					),
+				)
+				return
+			}
+
+			if (!upload(serverSocket)) {
+				statusCallback.accept(
+					UpdateStatusEvent(
+						deviceId,
+						FirmwareUpdateStatus.ERROR_UPLOAD_FAILED,
+					),
+				)
+				return
+			}
+
+			statusCallback.accept(
+				UpdateStatusEvent(
+					deviceId,
+					FirmwareUpdateStatus.REBOOTING,
+				),
+			)
+		}
+	}
+
+	companion object {
+		private const val FLASH = 0
+		private const val PORT = 8266
```

Comment:
Is ESP32 not planned for support here? ESP32 uses port `3232` for OTA but pretty much everything else is the same iirc.